### ISSUE TITLE
ENH: np.default_array_ufunc

### ISSUE THREAD

**Issue by jbrockmendel** [numpy/numpy#20044]
### Proposed new feature or change:

Context: I'm trying to implement `pandas.Index.__array_ufunc__` and `pandas.core.arrays.ExtensionArray.__array_ufunc__`

It would be helpful if there were a `np.default_array_ufunc` such that:

```
def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
     return np.default_array_ufunc(self, ufunc, method, *inputs, **kwargs)
```

would be equivalent to not implementing `__array_ufunc__` at all.

This would make it feasible to implement `__array_ufunc__` incrementally, in this case by using [maybe_dispatch_ufunc_to_dunder_op](https://github.com/pandas-dev/pandas/blob/master/pandas/_libs/ops_dispatch.pyx#L63) (a little bit like a reversed `NDArrayOperatorsMixin`) 

```
def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
    result = maybe_dispatch_ufunc_to_dunder_op(self, ufunc, method, *inputs, **kwargs)
    if result is not NotImplemented:
        return result
    return np.default_array_ufunc(self, ufunc, method, *inputs, **kwargs)
```

My not-quite-right implementation reads

```
def default_array_ufunc(self, ufunc: np.ufunc, method: str, *inputs, **kwargs):
    """
    Fallback to the behavior we would get if we did not define __array_ufunc__.

    Notes
    -----
    We are assuming that `self` is among `inputs`.
    """
    if not any(x is self for x in inputs):
        raise NotImplementedError

    new_inputs = [x if x is not self else np.asarray(x) for x in inputs]

    return getattr(ufunc, method)(*new_inputs, **kwargs)
```

This doesn't _quite_ achieve the goal of behaving as if `__array_ufunc__` were not implemented.  If it did, I think it would be better to live in numpy than in pandas.

xref #18186 motivated by trying to implement `__array_function__` incrementally.

**Comment by shoyer** [comment]
> This doesn't _quite_ achieve the goal of behaving as if `__array_ufunc__` were not implemented. If it did, I think it would be better to live in numpy than in pandas.

I'm curious, how is this implementation incorrect? At the very least, I think this is very close to being what you would want gere.

**Comment by jbrockmendel** [comment]
> Would you also have to do this for outputs in principle? `__array_wrap__` and `__array_prepare__` will not be called for your object, although you can fix that.

Yes.  So far the [default_array_ufunc](https://github.com/pandas-dev/pandas/blob/master/pandas/core/arraylike.py#L459) that I've used for [Index](https://github.com/pandas-dev/pandas/blob/master/pandas/core/indexes/base.py#L849) and [ExtensionArray](https://github.com/pandas-dev/pandas/blob/master/pandas/core/arrays/base.py#L1387) does this unwrapping for `*inputs` but not `**kwargs`.  The whole thing is patchwork and could use some expert eyeballs.

> For `__array_function__` I am a slightly unsure, because I am not sure that the fact that NumPy is not special is a good design (for performance reasons mainly).

FWIW my attempt to implement `__array_function__` stalled out largely bc tracking down where `self` is located among `args` and `**kwargs` was a rabbit hole xref #18186

**Comment by seberg** [comment]
One question I have here, is what would happen if you just use `_implementation()`?  Otherwise, I think we need to make a clearer proposal of what we could do in NumPy.

**Comment by jbrockmendel** [comment]
I'm not aware of _implementation.  should i just grep for that and try to implement something similar?

**Comment by seberg** [comment]
`_implementation` is private, but its the undispatched NumPy code, and I think it might be OK to consider it public.  The issue is, that _inside_ the function you would end up dispatching again probably.

**Comment by rgommers** [comment]
```
In [9]: np.mean._implementation
Out[9]: <function numpy.core.fromnumeric.mean(a, axis=None, dtype=None, out=None, keepdims=<no value>, *, where=<no value>)>
```

**Comment by jbrockmendel** [comment]
So is the suggestion that the relevant downstream class would implement `_implementation(?)` instead of `__array_ufunc__`?
